src/shipit_uplift/shipit_uplift/coverage.py

```diff
@@ -200,48 +210,85 @@ def get_latest_build():
         assert False, 'Not implemented'
 
     @staticmethod
-    @lru_cache(maxsize=32)
+    @lru_cache(maxsize=2048)
     def get_directory_coverage(changeset, prev_changeset, directory):
         assert False, 'Not implemented'
 
 
 coverage_service = CodecovCoverage()
 
 
-@lru_cache(maxsize=32)
-def get_push_range(push_id):
-    r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?tipsonly=1&startID=%s&endID=%s' % (push_id - 1, push_id + 7))
+# Push ID to build changeset
+MAX_PUSHES = 2048
+push_to_changeset_cache = LRUCache(maxsize=MAX_PUSHES)
+# Changeset to changeset data (files and push ID)
+MAX_CHANGESETS = 2048
+changeset_cache = LRUCache(maxsize=MAX_CHANGESETS)
+
+
+def get_pushes(push_id):
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?version=2&full=1&startID=%s&endID=%s' % (push_id - 1, push_id + 7))
     data = r.json()
-    return [(pushid, pushdata['changesets'][0]) for pushid, pushdata in data.items()]
+
+    for pushid, pushdata in data['pushes'].items():
+        pushid = int(pushid)
+
+        push_to_changeset_cache[pushid] = pushdata['changesets'][-1]['node']
+
+        for changeset in pushdata['changesets']:
+            is_merge = any(text in changeset['desc'] for text in ['r=merge', 'a=merge'])
+
+            if not is_merge:
```

Comment:
Nit: i would write first the "positive" case, then the other one. Feel free to ignore.